warehouse/multitenant/manager.go

```diff
@@ -0,0 +1,139 @@
+package multitenant
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/rudderlabs/rudder-server/config"
+	backendconfig "github.com/rudderlabs/rudder-server/config/backend-config"
+)
+
+var degradedWorkspaceIDs []string
+
+func init() {
+	config.RegisterStringSliceConfigVariable(nil, &degradedWorkspaceIDs, false, "Warehouse.degradedWorkspaceIDs")
+}
+
+type Manager struct {
+	BackendConfig        backendconfig.BackendConfig
+	DegradedWorkspaceIDs []string
+
+	sourceIDToWorkspaceID map[string]string
+	excludeWorkspaceIDMap map[string]struct{}
+
+	ready     chan struct{}
+	sourceMu  sync.Mutex
+	readyOnce sync.Once
+	initOnce  sync.Once
+}
+
+func (m *Manager) init() {
+	m.initOnce.Do(func() {
+		if m.DegradedWorkspaceIDs == nil {
+			m.DegradedWorkspaceIDs = degradedWorkspaceIDs
+		}
+
+		m.sourceIDToWorkspaceID = make(map[string]string)
+		m.excludeWorkspaceIDMap = make(map[string]struct{})
+
+		for _, workspaceID := range m.DegradedWorkspaceIDs {
+			m.excludeWorkspaceIDMap[workspaceID] = struct{}{}
+		}
+		m.ready = make(chan struct{})
+	})
+}
+
+// Run is a blocking function that executes manager background logic.
+func (m *Manager) Run(ctx context.Context) error {
+	m.init()
+
+	chIn := m.BackendConfig.Subscribe(ctx, backendconfig.TopicBackendConfig)
+	for data := range chIn {
+		m.sourceMu.Lock()
+		config := data.Data.(map[string]backendconfig.ConfigT)
+		for workspaceID := range config {
+			for _, source := range config[workspaceID].Sources {
+				m.sourceIDToWorkspaceID[source.ID] = workspaceID
+			}
+		}
+		m.sourceMu.Unlock()
+		m.readyOnce.Do(func() {
+			close(m.ready)
+		})
+	}
+
+	return nil
+}
+
+// DegradedWorkspace returns true if the workspaceID is degraded.
+func (m *Manager) DegradedWorkspace(workspaceID string) bool {
+	m.init()
+
+	for _, degradedWorkspaceID := range m.DegradedWorkspaceIDs {
+		if workspaceID == degradedWorkspaceID {
+			return true
+		}
+	}
+	return false
+}
+
+// DegradedWorkspaceIDs returns a list of degraded workspaceIDs.
+func (m *Manager) DegradedWorkspaces() []string {
+	m.init()
+
+	return m.DegradedWorkspaceIDs
+}
+
+// SourceToWorkspace returns the workspaceID for a given sourceID, even if workspaceID is degraded.
+// An error is returned if the sourceID is not found, or context is canceled.
+//
+//	NOTE: This function blocks until the backend config is loaded.
+func (m *Manager) SourceToWorkspace(ctx context.Context, sourceID string) (string, error) {
+	m.init()
+
+	select {
```

Comment:
Should we add a timeout here as well? Since these are being used in sources pending events.

Reply:
I believe `ctx` should be enough, the call can chose to provide a timeout or not.

In case for sousrce pending events endpoint, it will cancel when the HTTP requests gets canceled or timeouts

Reply:
My bad. Didn't see the request context being passed. Please ignore.